File: frontend/src/utils/markdownTable.ts
/**
 * Markdown 表格轻量解析工具
 *
 * 用于发布详情页只读渲染 release_doc（Markdown 2 列表格）。
 * 支持多行单元格：值中含换行时，后续不以 | 开头的行视为上一行值的延续。
 */

export interface MdTableRow {
  key: string;
  value: string;
}

/**
 * 解析 Markdown 2 列表格为行数组
 *
 * 仅解析 `| key | value |` 格式，跳过表头行和分隔行。
 * 支持多行单元格：当某行不以 | 结尾时，后续不以 | 开头的行
 * 视为该行 value 的延续行，用 \n 拼接。
 *
 * @param md Markdown 字符串
 * @returns 行数组
 */
export function parseMdTable(md: string): MdTableRow[] {
  if (!md) return [];
  const rows: MdTableRow[] = [];
  const lines = md.trim().split('\n');
  let i = 0;
  while (i < lines.length) {
    const trimmed = lines[i].trim();
    if (!trimmed.startsWith('|')) {
      i++;
      continue;
    }

    if (trimmed.endsWith('|')) {
      // 完整单行：| key | value |
      const cells = trimmed.split('|').map((c) => c.trim());
      const inner = cells.slice(1, -1);
      if (inner.length < 2) { i++; continue; }
      if (/^[-:\s]+$/.test(inner[0])) { i++; continue; }
      if ((inner[0] === '项目' || inner[0] === '字段') && inner[1] === '内容') { i++; continue; }
      rows.push({ key: inner[0], value: inner[1] });
      i++;
    } else {
      // 多行单元格：| key | value line 1 (未闭合)
      const cells = trimmed.split('|').map((c) => c.trim());
      const key = cells[1] || '';
      // key 之后到行尾的内容是 value 的第一行
      let valueParts = cells.slice(2).join('|').trim();
      i++;
      // 后续不以 | 开头的行都是 value 的延续
      while (i < lines.length && !lines[i].trim().startsWith('|')) {
        valueParts += '\n' + lines[i].trim().replace(/\|$/, '').trim();
        i++;
      }
      if (/^[-:\s]+$/.test(key)) continue;
      if ((key === '项目' || key === '字段') && valueParts === '内容') continue;
      rows.push({ key, value: valueParts });
    }
  }
  return rows;
}

/**
 * 将行数组序列化为 Markdown 2 列表格字符串
 *
 * 保持多行值原样输出（值中含 \n 时生成多行单元格），
 * 不拆分为多行也不使用 <br> 标签。
 *
 * 注意：parseMdTable -> buildMdTable 为幂等往返--含换行的值
 * 经往返后保持为同一行的多行单元格。
 *
 * @param rows 行数组
 * @returns Markdown 表格字符串（含最小表头以兼容 MD 语法）
 */
export function buildMdTable(rows: MdTableRow[]): string {
  const lines = ['| 字段 | 内容 |', '|------|------|'];
  for (const row of rows) {
    const safeKey = row.key.replace(/\|/g, '\\|');
    const safeValue = row.value.replace(/\|/g, '\\|');
    lines.push(`| ${safeKey} | ${safeValue} |`);
  }
  return lines.join('\n');
}

File: backend/utils/markdown_table.py
"""Markdown 表格工具

发布说明（release_doc）以 2 列表格存储，多行单元格使用字面换行；
标准 Markdown 表格不支持单元格内换行，导出场景（下载 MD / 推 SVN）
需要把单元格内的字面换行转换为 <br>。
"""


def table_newlines_to_br(md: str) -> str:
    """把 2 列表格多行单元格内的字面换行转换为 <br>。

    存储格式中，多行单元格跨越多个物理行：
        | 变更内容 | 第一行
        第二行
        第三行 |
    转换后合并为单行、以 <br> 连接：
        | 变更内容 | 第一行<br>第二行<br>第三行 |

    完整单行（| 开头且 | 结尾）与非表格内容原样保留。

    Args:
        md: 原始 Markdown 字符串

    Returns:
        转换后的 Markdown 字符串
    """
    if not md or "\n" not in md:
        return md

    lines = md.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    out: list[str] = []
    i = 0
    while i < len(lines):
        line = lines[i]
        stripped = line.strip()
        # 多行单元格起始行：以 | 开头但不以 | 结尾
        if stripped.startswith("|") and not stripped.endswith("|"):
            parts = [line.rstrip()]
            i += 1
            closed = False
            while i < len(lines):
                cont = lines[i].strip()
                if cont.startswith("|"):
                    break
                i += 1
                if cont.endswith("|"):
                    cont = cont[:-1].rstrip()
                    closed = True
                if cont:
                    parts.append(cont)
                if closed:
                    break
            if closed or len(parts) > 1:
                merged = "<br>".join(parts)
                out.append(merged + (" |" if closed else ""))
            else:
                out.append(line)
        else:
            out.append(line)
            i += 1
    return "\n".join(out)
